### PR TITLE
Implement FFT

### PR DESCRIPTION
**Pull request by unnonouno**
I implemented back-propable FFT.
This interface is different from that of `xp.fft` because Variable does not support complex.

### CODE REVIEW

File: chainer/functions/__init__.py [hunk 1]
```diff
@@ -194,6 +194,9 @@
 from chainer.functions.math.exponential import Log2  # NOQA
 from chainer.functions.math.exponential_m1 import expm1  # NOQA
 from chainer.functions.math.exponential_m1 import Expm1  # NOQA
+from chainer.functions.math.fft import fft  # NOQA
+from chainer.functions.math.fft import FFT  # NOQA
```
Comment (okuta) [inline chainer/functions/__init__.py]: Is this needed?

Comment (unnonouno) [reply chainer/functions/__init__.py]: It looks other FunctionNode classes are defined in this file. Is it better to remove these classes?

Comment (okuta) [reply chainer/functions/__init__.py]: Yes. Existing definitions are for backward compatibility.

File: chainer/functions/math/fft.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+from chainer import cuda
+from chainer import function_node
+from chainer.utils import type_check
+
+
+class FFT(function_node.FunctionNode):
+
+    """Fast Fourie transform."""
```
Comment (tkerola) [inline chainer/functions/math/fft.py]: Please correct the spelling to Fourier.

File: chainer/functions/math/fft.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        return real_y, imag_y
+
+    def backward(self, inputs, grads):
+        gr, gi = grads
+        gxi, gxr = FFT(self._method).apply((gi, gr))
+        return gxr, gxi
+
+
+def fft(real, imag):
+    """Fast Fourie transform.
```
Comment (tkerola) [inline chainer/functions/math/fft.py]: Please correct the spelling to Fourier.

File: chainer/functions/math/fft.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    Returns:
+        tuple: Returns ``(ry, ri)`` where ``ry`` is the real part of
+        the result and ``ri`` is the imaginary part of the result.
+
+    """
+    return FFT('fft').apply((real, imag))
+
+
+def ifft(real, imag):
+    """Inverse fast Fourie transform.
```
Comment (tkerola) [inline chainer/functions/math/fft.py]: Please correct the spelling to Fourier.

**Comment by unnonouno** [comment]
We discussed about its interface on offline, and decided to use `tuple` now. After Chainer supports complex numbers, we also support them by checking input type.

**Comment by okuta** [comment]
flake8
```
./chainer/functions/math/fft.py:44:80: E501 line too long (80 > 79 characters)
./chainer/functions/math/fft.py:65:80: E501 line too long (80 > 79 characters)
```

And, why do not check double-backward?

**Comment by unnonouno** [comment]
fixed

**Comment by unnonouno** [comment]
fixed

**Comment by okuta** [comment]
Please add `fft` and `ifft` to rst file.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**